rbd/rbd.go

```diff
@@ -1215,3 +1244,63 @@ func TrashRestore(ioctx *rados.IOContext, id, name string) error {
 
 	return GetError(C.rbd_trash_restore(C.rados_ioctx_t(ioctx.Pointer()), c_id, c_name))
 }
+
+// OpenImage will open an existing rbd image by name and snapshot name, returning
+// a new opened image.
+//
+// Implements:
+//  int rbd_open(rados_ioctx_t io, const char *name,
+//               rbd_image_t *image, const char *snap_name);
+func OpenImage(ioctx *rados.IOContext, name, snapName string) (*Image, error) {
```

Comment:
I would prefer a clear distinction between OpenImage() and OpenSnapshot(). The passed `cSnapName` needs to be `NULL` or a string according to librbd.h. Passing an empty string `""` may not be the same as passing `NULL`.

Reply:
I have to think about this a little bit. I'm torn between not wanting to deviate too much from the underlying APIs and the more "typed" variation you suggest.

FWIW I did inspect the code in ceph and do a few tests and generally passing a null string seems equivalent to empty string in most cases. However, I'm open to the suggestion to do the safer option of sticking to null. But we can also change the empty string in go to a nil in go code before calling the c api, without changing the function call in go.

Yet another option would be to have a special type for the snap name and a sentinel value that gets converted to a null string. Something like:
```go

type SnapshotName string
var NoSnapshot = SnapshotName("")

func OpenImage(ioctx *rados.IOContext, name string, snapName SnapshotName) (*Image, error) {
    // ....
   if snapName == NoSnapshot {
       cSnapName = C.String(null)
   } else {
      cSnapName = C.String(string(snapName))
  }
  // ...
```

I'll meditate on it a bit more.

Reply:
I tried to make a compromise between deviating from the ceph api pattern but being more explicit about opening a non-snapshot image. Please take a look at the final patch in the series for that and let me know what you think.

Reply:
I strongly prefer `OpenImage()` and `OpenImageReadOnly()` over the `GetImage()` approach. Ideally `GetImage()` gets deprecated too. No need to `OpenReadWrite()` and `OpenReadOnly()` in that case.

I like the passing of such a `NoSnapshot` argument that gets translated to `nil` as argument in `rbd_open()` (not a `C.String(null)` (I dont know what that translates to). It is a little more Golang native, and still follows the librbd API closely.

Reply:
> I strongly prefer `OpenImage()` and `OpenImageReadOnly()` over the `GetImage()` approach. Ideally `GetImage()` gets deprecated too. No need to `OpenReadWrite()` and `OpenReadOnly()` in that case.

In general, I do too. However, I didn't want to force a major change in the current flow. Currently a lot of the api does:
```
img := Create(...)
img.Open()
```
But the underlying apis don't actually create any persistent objects for the create calls. Instead, it would work more like:
```
Create(name, ...)
img := OpenImage(name, ....)
```
So I created these "helper" functions to avoid changing the old workflow too much.
I'm not sure if you prefer to simply drop the new helpers and just continue to use the deprecated Open function or push more toward a better underlying api?
I also think we could consider a different Create-and-Open helper function but I was hesitant to try and design that w/o more experience with the ceph apis.

> 
> I like the passing of such a `NoSnapshot` argument that gets translated to `nil` as argument in `rbd_open()` (not a `C.String(null)` (I dont know what that translates to). It is a little more Golang native, and still follows the librbd API closely.

I'm not quite clear if this means you have an issue with the change or are happy with it.

Reply:
What I do not like, is the `GetImage()` or a `Create()` that does not correspond to something that librbd offers. So, I would like the only way to get an RbdImage, is to call `OpenImage()` or similar. That does mean `RemoveImage()` needs to be added as well (and maybe others).

This is a different workflow from what users currently follow. But if we want go-ceph to but close to the librbd API, we need to make a switch-over at one point. Deprecating the non-matching API is a step in that direction.

The `OpenImage()` example you posted above is something I agree with.

Reply:
Gotcha.
I see that the pattern for now will be something like:
```
_, err := CreateX(ioctx, name, ...)
img, err := OpenImage(ioctx, name, NoSnapshot)
```
We'll be discarding the image objects returned from create frequently but that seems reasonable enough. I'll prepare a demo patch later today to see how it flows.

Reply:
Well, not completely. `Create()` would not need to return an image, the corresponding librbd functions do not do that either. An image seems equivalent to a filedescriptor that has been opened. I do not know if opening a non-existing image can be used to create the image as well (like `open(..., O_CREAT)` does).

Reply:
Yes, that's the way it should have been, but the api returns `*Image` today. Taking that away from the existing functions would be a breaking change IMO. So for now I think we need to continue returning the same types from these functions. However, we could have variants of the  Create functions that do not return images, and consider the other ones deprecated as well.
Thoughts?

I looked at the ceph apis a few weeks ago for an atomic create-and-open and did not see one, but I'd be happy to be corrected if I overlooked it.

Reply:
Come to think of it, since Create4 is fairly new we could either change the function signature or create a variant that does not return *Image (call it CreateImage?) and call that the "official" api for creating a new rbd image as long as the RbdImageOptions can do everything the 1-3 calls do.

Reply:
Clone(...) is also a call that "falsely" returns images.

Reply:
> I looked at the ceph apis a few weeks ago for an atomic create-and-open and did not see one, but I'd be happy to be corrected if I overlooked it.

I can not find this either.

> Come to think of it, since Create4 is fairly new we could either change the function signature or create a variant that does not return *Image (call it CreateImage?) and call that the "official" api for creating a new rbd image as long as the RbdImageOptions can do everything the 1-3 calls do.

I like `CreateImage()` with `RbdImageOptions` and not returning an `Image`. We could even drop `Create4()` in favour of it.

> Clone(...) is also a call that "falsely" returns images.

With that, `CloneImage()` can accept `RbdImageOptions` too and call `rbd_clone3()`.

Reply:
Hehe. I was working on clone3 yesterday afternoon anticipating this response. :-)
I'll try to whip up some more patches for this PR for the new style... I think we're on the same page conceptually.